File: 2021/Python/Day03.py
def Run(filepath):
    with open(filepath) as input:
        diagnostics = list(map(str, input.read().split('\n')))

    print('Day 3')
    Part1(diagnostics)
    Part2(diagnostics)

def Part1(diagnostics):
    #Fill strings with the right number of characters, to be replaced as the bits are calculated below
    gamma = '#' * len(diagnostics[0])
    epsilon = '#' * len(diagnostics[0])

    for bit in range(len(diagnostics[0])):
        ones = zeroes = count = leastCommonBit = mostCommonBit = 0

        for line in diagnostics:
            #Just add the value in the 'bit' position. If it's 1 it will increment 'ones', if it's 0 it won't
            ones += int(line[bit])
            count += 1

        #Because there are only ones and zeroes: 
        zeroes = count - ones

        if ones > zeroes:
            mostCommonBit = 1
        elif zeroes > ones:        
            leastCommonBit = 1

        #Insert results into the relevant spot while keeping the rest of the string intact
        gamma = gamma[:bit] + str(mostCommonBit) + gamma[bit+1:]
        epsilon = epsilon[:bit] + str(leastCommonBit) + epsilon[bit+1:]

    #Convert strings into binary numbers, i.e base-2 integers
    gamma = int(gamma,2)
    epsilon = int(epsilon,2)
    powerConsumption = (gamma*epsilon)

    print('Part 1: ' + str(powerConsumption))
    #Result should be 3912944

def Part2(diagnostics):    
    def BitFrequency(diagnostics, position, task):
        ones = zeroes = count = mostCommonBit = 0
        for entry in diagnostics:
            #Just add the value in the relevant position. If it's 1 it will increment 'ones', if it's 0 it won't
            ones += int(entry[position])
            count +=1
        #Because there are only ones and zeroes: 
        zeroes = count - ones
        if zeroes > ones:
            mostCommonBit = 0
        #Because in a draw, 1 wins:
        else: mostCommonBit = 1
        if task == 'o2':
            return mostCommonBit
        elif task == 'co2':
            #Because in both possible cases, this will result in the least common bit:
            return (1 - mostCommonBit)

    def Highlander(diagnostics, task):
    #There can be only one... (item left in the list)
        for position in range(len(diagnostics[0])):
            #Get either the most or least common bit for the o2 and co2 tasks respectively
            comparisonBit = BitFrequency(diagnostics, position, task)
            #Keep only the entries which have comparisonBit in the current position:
            diagnostics = [entry for entry in diagnostics if int(entry[position]) == comparisonBit]
            #Stop when there's only one left
            if len(diagnostics) <= 1:
                break
        return diagnostics[0]

    #Get binary numbers made up of most (o2) and least (co2) common bits
    o2 = int(Highlander(diagnostics,'o2'),2)
    co2 = int(Highlander(diagnostics,'co2'),2)
    lifeSupport = o2 * co2

    print('Part 2: ' + str(lifeSupport))
    #Result should be 4996233

File: 2021/CSharp/App/Day03.cs
namespace AdventOfCode;
public class Day03 : Day{
    public override Int64 Part1(string filepath){
        string[] file = File.ReadAllLines(filepath);
        List<string> diagnostics = new List<string>(file);
        string gamma = "", epsilon = "";
        int position = 0;
        //Fill strings with the right number of characters, to be replaced as the bits are calculated below
        gamma = gamma.PadLeft(diagnostics[0].Length, '#');
        epsilon = epsilon.PadLeft(diagnostics[0].Length, '#');

        while (position < diagnostics[0].Length){
            var (mostCommonBit, leastCommonBit) = BitFrequency(diagnostics, position);
            gamma = gamma.Substring(0, position) + mostCommonBit + gamma.Substring(position + 1);
            epsilon = epsilon.Substring(0, position) + leastCommonBit + epsilon.Substring(position + 1);
            position++;
        }
        int powerConsumption = Convert.ToInt32(gamma, 2) * Convert.ToInt32(epsilon, 2);
        
        System.Console.WriteLine($"Day 3-1: {powerConsumption}");
        return powerConsumption;
    }
    public override Int64 Part2(string filepath){
        string[] file = File.ReadAllLines(filepath);
        List<string> diagnostics = new List<string>(file);
        int o2 = CalculateLifeSupport(diagnostics, "o2");
        int co2 = CalculateLifeSupport(diagnostics, "co2");
        int lifeSupport = o2 * co2;
        
        System.Console.WriteLine($"Day 3-2: {lifeSupport}");
        return lifeSupport;
    }
    static int CalculateLifeSupport(List<string> diagnostics, string task){
        string comparisonBit;

        for (int position = 0; position < diagnostics[0].Length && diagnostics.Count > 1; position++){
            var (mostCommonBit, leastCommonBit) = BitFrequency(diagnostics, position);
            switch (task){
                case "co2":
                    comparisonBit = leastCommonBit;
                    break;
                default:
                    comparisonBit = mostCommonBit;
                    break;
            }
            diagnostics = diagnostics.Where(x => Convert.ToString(x[position]) == comparisonBit).ToList();
        }
        int rating = Convert.ToInt32(diagnostics[0], 2);
        return rating;
    }
    static (string most, string least) BitFrequency(List<string> diagnostics, int position){
        //Calculate the most and least common bit in the current postion in all strings
        int count = 0, ones = 0;
        string mostCommonBit, leastCommonBit;

        foreach (string diagnostic in diagnostics){
            int bit = int.Parse(diagnostic[position].ToString());
            //Just adds the bit rather than checking if it's a 1
            // because if it's a 0 that won't affect the value
            ones += bit;
            count++;
        }
        if (ones < count - ones){ //count-ones = the number of 0s in the string
            mostCommonBit = "0";
            leastCommonBit = "1";
        }
        else{mostCommonBit = "1";
            leastCommonBit = "0";
        }
        return (mostCommonBit, leastCommonBit);
    }
}